Api

Aidaho12 opened the pull request.

Aidaho12 committed:
Refactor authentication and improve error handlers

This commit encompasses a series of improvements. Authentication mechanism has been shifted from Flask-Login to Flask-JWT for enhanced security. Changes include updates in request routes and functions across multiple files to accommodate this switch. Unnecessary import statements have also been cleaned up. Furthermore, robust error handling has been implemented to increase application resilience and stability. To maintain compliance, commented out codes and non-used functions have been removed. Lastly, variables and functions have been renamed to improve code readability and maintainability.

Aidaho12 committed:
Add ServerView and GroupView, modify group updating and deletion

Introduced a ServerView class and GroupView class providing POST, PUT and DELETE methods handling operations for servers and groups. Modified the 'update_group' and 'delete_group' in group.py to raise exceptions and enhance error reporting visibility. These changes provide better management of servers and groups.

Aidaho12 committed:
Update SSH credentials management and enhance exception handling

Changes include additon of handling for additional exceptions in SSH credentials management. This is complemented by improved type checking for group affiliation and roles. This improves robustness and security by ensuring that invalid or inappropriate actions are more efficiently detected and addressed.

StuBz211 commented on app/__init__.py:
```diff
@@ -1,22 +1,29 @@
+from app.api.v1.routes.user import bp as user_api_bp
+from app.api.v1.routes.server import bp as server_api_bp
+
+app.register_blueprint(main_api_bp, url_prefix="/api/v1.0")
```
main_api_bp.regisre(UserBP, serverBP)

StuBz211 commented on app/api/v1/routes/main/routes.py:
```diff
@@ -0,0 +1,50 @@
+    try:
+        login = request.json.get('login')
+        password = request.json.get('password')
+    except Exception:
```
проверить возможные ошибки, может не потребуется.

Exception as e:
print(e)

StuBz211 commented on app/api/v1/routes/main/routes.py:
```diff
@@ -0,0 +1,50 @@
+    try:
+        user_params = roxywi_auth.check_user_password(login, password)
+    except Exception as e:
+        return roxywi_common.handle_json_exceptions(e, ''), 200
```
!=200

StuBz211 commented on app/api/v1/routes/server/routes.py:
```diff
@@ -0,0 +1,7 @@
+from app.views.server.views import ServerView, GroupView, CredsView
+
+
+bp.add_url_rule('', view_func=ServerView.as_view('server', True))
```
```# TODO rewrite as api with serializers```

StuBz211 commented on app/config.py:
```diff
@@ -1,7 +1,11 @@
     CACHE_TYPE = 'SimpleCache'
     CACHE_DEFAULT_TIMEOUT = 3000
     SCHEDULER_API_ENABLED = True
+    JWT_SECRET_KEY = "very secret salt to protect your RMON jwt"
```
os.evniron?

StuBz211 commented on app/login.py:
```diff
@@ -1,20 +1,20 @@
             'login_page', 'static', 'main.show_roxywi_version', 'smon.show_smon_status_page', 'smon.smon_history_statuses_avg',
-            'smon.smon_history_statuses', 'smon.agent_get_checks', 'smon.get_check_status', 'smon.smon_history_metric'
-    ):
+            'smon.smon_history_statuses', 'smon.agent_get_checks', 'smon.get_check_status', 'smon.smon_history_metric', 'api'
```
что это такое? 
`('smon.smon_history_statuses', 'smon.agent_get_checks', 'smon.get_check_status', 'smon.smon_history_metric', 'api'`

Это эндпоинты в которых надо проверить авторизацию?

может стоит вынести в переменную? 
типо allowed_endpoins ?

StuBz211 commented on app/login.py:
```diff
@@ -30,75 +30,47 @@ def check_login():
-        users = user_sql.select_users(user=login)
+    if request.method == 'GET':
+        try:
+            lang = roxywi_common.get_user_lang_for_flask()
```
можно вшить дефолнтый язык в самой функции и там же отлавливать ошибку.
`roxywi_common.get_user_lang_for_flask()`

тогда не придется отлавливать ошибки снаружи.

StuBz211 commented on app/modules/roxywi/auth.py:
```diff
@@ -127,8 +129,33 @@ def do_login(user_uuid: str, user_group: str, user: str, next_url: str):
+
+        for user in users:
+            if user.activeuser == 0:
+                raise Exception('Your login is disabled')
```
заведи собственные ошибки
 
raise AuthDisabledErrot

например 
AuthDisabledErrot, BanError

StuBz211 commented on app/modules/roxywi/auth.py:
```diff
@@ -127,8 +129,33 @@ def do_login(user_uuid: str, user_group: str, user: str, next_url: str):
+
+
+def check_user_password(login: str, password: str) -> dict:
+    if login and password:
```
можно избавить от дополнительной вложености 

if not login or not password:
    raise AuthCredentialsError 
    
  статус ответа должен быть 400 в таком случае

StuBz211 commented on app/views/user/views.py:
```diff
@@ -0,0 +1,321 @@
+        """
+        try:
+            user_id = int(self.json_data["user_id"])
+        except ValueError as e:
```
Если хочешь обработать сразу все ошибки то можно передать кортеж

except (ValueError, KeyError) as err: 
     handle_error(err)

StuBz211 replied:
?
```
try:
   int( ....
except Exception as err: 
    return  handle_error(err)
    
 def handle_error(err) 
       if isisnance(err, (KeyError, ValueError):
            return  roxywi_common.handle_json_exceptions(e, f'There is must be a value')
```

StuBz211 commented on app/views/user/views.py:
```diff
@@ -0,0 +1,321 @@
+        try:
+            user = user_sql.get_user_id(user_id)
+        except Exception as e:
+            return roxywi_common.handle_json_exceptions(e, 'Cannot get user')
```
roxywi_common.handle_json_exceptions(e, 'Cannot get user'),   404 ? or 500 ?

StuBz211 commented on app/modules/roxywi/common.py:
```diff
@@ -15,15 +17,13 @@
+		claims = get_jwt()
+		user_group_id = claims['group']
 		groups = group_sql.select_groups(id=user_group_id)
 		for g in groups:
```
```for g in groups:``` -> ```for group in groups:```

глобально у тебя определена `g` в цикле её можно заместить.

StuBz211 commented on app/middleware.py:
```diff
@@ -1,20 +1,47 @@
+        @wraps(fn)
+        def decorated_views(*args, **kwargs):
+            if not roxywi_auth.is_admin(level=level):
+                return abort(400, 'bad permission')
```
403 ?

StuBz211 commented on app/modules/roxywi/auth.py:
```diff
@@ -100,25 +104,23 @@ def create_uuid_and_token(login: str):
-
-    if not session_ttl:
-        session_ttl = 5
+    # try:
```
token ttl, in readme for customers?!

StuBz211 commented on app/modules/roxywi/auth.py:
```diff
@@ -100,25 +104,23 @@ def create_uuid_and_token(login: str):
-    resp.set_cookie('uuid', user_uuid, secure=True, expires=expires.strftime("%a, %d %b %Y %H:%M:%S GMT"), samesite='Strict')
-    resp.set_cookie('group', str(user_group), expires=expires.strftime("%a, %d %b %Y %H:%M:%S GMT"), samesite='Strict')
+    response = jsonify({"status": "done", "next_url": redirect_to})
+    additional_claims = {'uuid': user_uuid, 'group': str(user_group)}
```
нужн ли user_uuid ?

StuBz211 commented on app/views/user/views.py:
```diff
@@ -0,0 +1,321 @@
+        try:
+            group_id = int(self.json_data["group_id"])
+        except ValueError as e:
+            return roxywi_common.handle_json_exceptions(e, f'There is must be a value')
```
`group_id` is must be a int ?

Aidaho12 committed:
Refactor code for user authentication and exception handling

The code has been refactored to improve user authentication process and exception handling. The checks for user login and password are enhanced for better security. Unnecessary checks have been removed to simplify the code. The exception handling is improved by introducing a dedicated function to handle exceptions related to JSON data and code is restructured to cleanly handle individual types of exceptions. The API routes have been adjusted for better structure and consistency. User deletion function now does not return any value.

Aidaho12 committed:
Add error handling and improve API endpoints

Refactor the code to improve API endpoints and change the way server responses are handled in ajax requests. Improve the user experience by providing direct and specific error responses across all endpoints and extracting the error messages centrally. Rename 'activeuser' to 'enabled' in multiple locations for code clarity and consistency.

StuBz211 commented on app/modules/db/db_model.py:
```diff
@@ -183,10 +182,10 @@ class Meta:
     username = CharField()
     password = CharField(null=True)
-    groups = IntegerField(constraints=[SQL('DEFAULT 1')])
+    group_id = IntegerField(constraints=[SQL('DEFAULT 1')])
```
IntegerField(default=1)  - ? либа peeweee, допускает подобный синтаксис?

StuBz211 replied:
Это точно след ПР такое делать.

Aidaho12 replied:
Вроде умеет, попробую в следующем ПР

StuBz211 commented on app/modules/roxywi/class_models.py:
```diff
@@ -0,0 +1,143 @@
+            try:
+                timeout = int(values['timeout'])
+            except ValueError:
+                raise RoxywiValidationError('timeout must be an integer')
```
здесь скорее уместно использовать дефолтную ошибки валидации от пидантика.

from pydantic import ValidationError

Aidaho12 committed:
Refactor database handling for group-related functionality

This commit refactors how the application handles database interactions for group-related functionality. Notably, it introduces foreign key constraints for more robust data integrity, replaces a CharField with a ForeignKeyField to establish relationships between tables, and updates methods to accommodate these changes. It also removes unused code blocks to clean up the codebase. Additionally, some error handling functionality has been improved.
